include/boost/regex/v4/u32regex_iterator.hpp

```diff
@@ -72,21 +72,13 @@ class u32regex_iterator_implementation
 
 template <class BidirectionalIterator>
 class u32regex_iterator 
-#ifndef BOOST_NO_STD_ITERATOR
-   : public std::iterator<
-         std::forward_iterator_tag, 
-         match_results<BidirectionalIterator>,
-         typename BOOST_REGEX_DETAIL_NS::regex_iterator_traits<BidirectionalIterator>::difference_type,
-         const match_results<BidirectionalIterator>*,
-         const match_results<BidirectionalIterator>& >         
-#endif
 {
 private:
    typedef u32regex_iterator_implementation<BidirectionalIterator> impl;
    typedef shared_ptr<impl> pimpl;
 public:
    typedef          u32regex                                                regex_type;
-   typedef          match_results<BidirectionalIterator>                    value_type;
```

Comment:
`typename` is wrong here. Not sure why you decided to add it.

Reply:
I see ☹️ . My problem is that the very same test-cases failed without 'typename' on travis. Starting without that 'typename' I've got lengthy error messages about a missing 'value_type'. Reluctantly adding it (I saw no obvious reason for that) is giving much short error messages - well, at least some sort of improvement. Unfortunately I have no other means of running tests with gcc or clang on a sufficiently matching setup.

Reply:
It's too convoluted to follow and I get no failures with mingw g++ 7, I guess John has to take a look at it. `match_results` is for some reason declared in `v4/regex_format.hpp:56`, which I can only suspect is not being included prior to this header, hence the errors. But it's hard to tell. Either way, I think that it's more correct without the `typename` than with it.

Reply:
Ah, there are no errors because `icu.cpp` only includes `icu.hpp` when `BOOST_HAS_ICU` is defined and it isn't for me.

Reply:
I suspect that std::iterator_traits is implemented SFINAE-friendly, and the ```value_type``` member is no longer present then. I've reworked the patch such that it should no longer SFINAE-away those members now.

Reply:
Installing libicu-devel under Cygwin reproduces the errors.

Reply:
BTW, it looks like Boost.Regex isn't ICU-enabled for all Windows test-runs in Boost's test-matrix. I had to spend quite some time to make this work here on my local Windows machine. Some tests for the ```wchar_t```-overloads were failing. It seems like there is an implicit assumption that ICU's ```UChar``` is the same as ```wchar_t```. But it's not: ```UChar``` is typedef'd to ```char16_t``` and that' *not* the same as ```wchar_t```.